srml/im-online/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+			}
+		}
+
+		// Runs after every block.
+		fn offchain_worker(block_number: T::BlockNumber) {
+			fn gossip<T: Trait>(block_number: T::BlockNumber) -> Result<(), OffchainErr> {
+				match sr_io::local_authority_pubkey(<T as Trait>::CRYPTO_KIND) {
+					Ok(key) => {
+						let authority_id = <T as Trait>::AuthorityId::decode(&mut &key[..])
+							.ok_or(OffchainErr::DecodeAuthorityId)?;
+						let network_state =
+							sr_io::local_network_state().map_err(|_| OffchainErr::LocalNetworkState)?;
+						let heartbeat_data = Heartbeat {
+							block_number,
+							network_state,
+							session_index: <session::Module<T>>::current_index(),
+							authority_id,
+						};
+
+						match sr_io::sign(None, <T as Trait>::CRYPTO_KIND, &heartbeat_data.encode()) {
+							Ok(signature) => {
+								let call = Call::heartbeat(heartbeat_data, signature);
+								let ex = T::UncheckedExtrinsic::new_unsigned(call.into())
+									.ok_or(OffchainErr::ExtrinsicCreation)?;
+								sr_io::submit_transaction(&ex)
+									.map_err(|_| OffchainErr::SubmitTransaction)?;
+								Ok(())
+							},
+							Err(_) => Err(OffchainErr::FailedSigning),
+						}
+					},
+					Err(_) => {
+						// we only run when a local authority key is configured
+						Ok(())
+					}
+				}
+			}
+
+			let gossip_at = <GossipAt<T>>::get();
+			let now = <system::Module<T>>::block_number();
+			if gossip_at == now {
+				match gossip::<T>(block_number) {
+					Ok(_) => {},
+					Err(_) => {
+						print("Error in offchain worker!");
+					},
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Returns `true` if a heartbeat has been received for `AuthorityId`
+	/// during the current era. Otherwise `false`.
+	pub fn is_online_in_current_era(authority_id: &T::AuthorityId) -> bool {
+		<ReceivedHeartbeats<T>>::exists(PREFIX, authority_id)
+	}
+
+	/// Session has just changed.
+	fn new_session() {
+		let now = <system::Module<T>>::block_number();
+		<GossipAt<T>>::put(now);
+
+		let current_session = <session::Module<T>>::current_index();
+
+		match LastNewEraStart::get() {
+			Some(last_new_era_start) => {
+				let sessions_per_era = T::SessionsPerEra::get();
+
+				// clear map after each era
+				let new_era = current_session - last_new_era_start > sessions_per_era;
+				if new_era {
+					LastNewEraStart::put(current_session);
+					<ReceivedHeartbeats<T>>::remove_prefix(PREFIX);
+				}
+			},
+			None => LastNewEraStart::put(current_session),
+		};
+	}
+}
+
+impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
+	type Key = <T as Trait>::AuthorityId;
+
+	fn on_new_session<'a, I: 'a>(_changed: bool, _validators: I) {
+		Self::new_session();
+	}
+
+	fn on_disabled(_i: usize) {
+		// ignore
+	}
+}
+
+impl<T: Trait> srml_support::unsigned::ValidateUnsigned for Module<T> {
+	type Call = Call<T>;
+
+	fn validate_unsigned(call: &Self::Call) -> srml_support::unsigned::TransactionValidity {
+		match call {
+			Call::heartbeat(heartbeat, signature) => {
```

Comment:
We need additional check to make sure the `hearbeat` was not already sent by that `authority_id`. Most likely using `is_online_in_current_era`